tools/build

```diff
@@ -0,0 +1,128 @@
+#!/bin/sh
+
+set -ex
+
+print() {
+    printf '%s\n' "$*"
+}
+
+die() {
+    st=${?:-0}
+    if [ $st -eq 0 ]; then
+        st=2
+    fi
+    print "$*" >&2
+    exit $st
+}
+
+usage() {
+    die 'build [ -n -pg1[234] ] ( test-crates | test-extension | test-post-install | test-doc | clippy )'
+}
+
+require_pgversion() {
+    [ -n "$pgversion" ] || die 'specify one of -pg12 | -pg13 | -pg14'
+}
+
+find_pgconfig() {
+    require_pgversion
+    full_version=$(
+        cd "$HOME/.pgx"
+        set -- $pgversion.*/pgx-install
+        [ "$1" = "$pgversion.*/pgx-install" ] && die "$pgversion not installed in $HOME/.pgx"
+        set -- $(dirname $*)
+        [ $# -eq 1 ] || die "too many installations for $pgversion, found: $*"
+        echo $1)
+    pg_config="$HOME/.pgx/$full_version/pgx-install/bin/pg_config"
+}
+
+[ $# -ge 1 ] || usage
+
+# For reasons we don't yet understand, pgx prevents the cargo cache from
+# working across runs UNLESS all pgx commands are run from the 'extension'
+# subdirectory of the project.  Any cargo command we need to run from that
+# subdirectory, we use CARGO_TARGET_DIR="$extension_CARGO_TARGET_DIR", and for
+# all others, CARGO_TARGET_DIR="$top_CARGO_TARGET_DIR".
+#
+# If each entry in ci.yml `jobs` is guaranteed to be run in isolation, always
+# with the target dir freshly copied from the specified cache key, then we
+# don't need to bother with CARGO_TARGET_DIR here.
+#
+# However, this script is derived from one I was using for my local builds
+# anyway, and apparently everyone is suffering the slow build times locally,
+# so why have more than one way to build?  Why not use the same tool locally
+# as the official build uses?
+#
+# CARGO_TARGET_DIR_NAME allows me to call my target directory '.target'.
+# TODO Can we simplify and just force .target here?  It works fine, but I didn't want to surprise people...
```

Comment:
do you mean `./target`? If so, I think that should be safe to use. If we really wanted to be principled, I guess we could try to extract `target_directory` it from `cargo metadata` 🤔

Reply:
(users _can_ override the default target directory using a [`.cargo/config.toml`](https://doc.rust-lang.org/cargo/reference/config.html))

Reply:
I mean `.target`. I always prefix my build directories with `.` so I don't trip over them all the time.

I experimented with the setting in `~/.cargo/config.toml` and it doesn't seem to be very useful: it overrides for ALL projects and JUST ONE DIRECTORY and resolves it relative to the home directory! Setting `build.target-dir = "target"` causes `/home/epg/target` to be used for all Rust builds which works about as well as you'd expect :)